File: 20-actors/niyaku/methods/agv_transfer.py
"""agv_transfer — automated guided vehicle (AGV) horizontal-transport core.

After the STS crane lands a box on the quay apron, a battery AGV carries it to
the yard stack. This module is the planning core behind the `yard_transfer` cell:

  * a **trapezoidal velocity profile** (accel → cruise → decel, or a triangular
    profile when the leg is too short to reach cruise) giving time-optimal,
    jerk-bounded travel time over a leg;
  * a **lane-segment conflict check** — two AGVs sharing a one-way segment must
    not have overlapping occupancy time windows (deadlock/collision avoidance);
  * a **greedy dispatch** that assigns moves to a fleet to minimise makespan.

stdlib-only · pywasm-ready. Pure planning compute; dispatches no real vehicle
(G12 no-server-key / G13 consent-bound).

Per ADR-2606082000 (niyaku R0).
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Dict, List, Tuple


@dataclass
class Agv:
    """Battery AGV kinematic envelope (electric, regenerative braking — G8)."""
    v_max: float = 6.0      # m/s
    a_max: float = 0.8      # m/s² (accel = decel)
    length_m: float = 16.0  # AGV + 40ft box footprint, for segment occupancy


def travel_time(distance_m: float, agv: Agv) -> float:
    """Time-optimal travel time over a straight leg under a trapezoidal profile.

    If the leg is long enough to reach ``v_max`` the profile is accel/cruise/
    decel; otherwise it is a symmetric triangular accel/decel that peaks below
    ``v_max``.
    """
    if distance_m < 0:
        raise ValueError("distance must be non-negative")
    if distance_m == 0:
        return 0.0
    a = agv.a_max
    v = agv.v_max
    d_to_vmax = v * v / a            # distance to accel to v_max then decel to 0
    if distance_m >= d_to_vmax:
        t_ramp = v / a               # accel + symmetric decel
        d_cruise = distance_m - d_to_vmax
        return 2.0 * t_ramp + d_cruise / v
    # triangular: peak velocity vp = sqrt(a * d); total t = 2 vp / a
    vp = math.sqrt(a * distance_m)
    return 2.0 * vp / a


@dataclass
class SegmentReservation:
    """One AGV's occupancy of a named one-way lane segment over [t_in, t_out]."""
    segment: str
    agv_id: str
    t_in: float
    t_out: float


def reservations_conflict(r1: SegmentReservation, r2: SegmentReservation) -> bool:
    """True iff two reservations share a segment and overlap in time.

    Touching at an endpoint (t_out == t_in) is NOT a conflict (one clears as the
    other enters). Different segments never conflict.
    """
    if r1.segment != r2.segment or r1.agv_id == r2.agv_id:
        return False
    return r1.t_in < r2.t_out and r2.t_in < r1.t_out


def find_conflicts(reservations: List[SegmentReservation]) -> List[Tuple[int, int]]:
    """All conflicting index pairs (i<j) in a reservation list."""
    out: List[Tuple[int, int]] = []
    for i in range(len(reservations)):
        for j in range(i + 1, len(reservations)):
            if reservations_conflict(reservations[i], reservations[j]):
                out.append((i, j))
    return out


@dataclass
class Move:
    move_id: str
    distance_m: float


@dataclass
class DispatchResult:
    # agv_id -> ordered list of move_ids
    assignment: Dict[str, List[str]] = field(default_factory=dict)
    # agv_id -> total busy time (s)
    finish_time: Dict[str, float] = field(default_factory=dict)

    def makespan(self) -> float:
        return max(self.finish_time.values()) if self.finish_time else 0.0


def dispatch(moves: List[Move], agv_ids: List[str], agv: Agv) -> DispatchResult:
    """Greedy makespan-minimising assignment: each move (longest-first) goes to
    the AGV that is currently free earliest (LPT — longest-processing-time rule).
    """
    if not agv_ids:
        raise ValueError("need at least one AGV")
    res = DispatchResult(
        assignment={a: [] for a in agv_ids},
        finish_time={a: 0.0 for a in agv_ids},
    )
    for mv in sorted(moves, key=lambda m: -m.distance_m):
        # pick the AGV that frees up soonest
        a = min(agv_ids, key=lambda x: res.finish_time[x])
        res.assignment[a].append(mv.move_id)
        res.finish_time[a] += travel_time(mv.distance_m, agv)
    return res

File: 20-actors/niyaku/methods/agv_transfer.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/agv_transfer.py.
(ns niyaku.methods.agv-transfer
  "agv_transfer — automated guided vehicle (AGV) horizontal-transport core (ADR-2606082000).

  After the STS crane lands a box on the quay apron, a battery AGV carries it to the yard
  stack. Planning core behind the yard_transfer cell:
    * a trapezoidal velocity profile (accel→cruise→decel, or triangular when the leg is too
      short to reach cruise) giving time-optimal travel time;
    * a lane-segment conflict check — two AGVs on a one-way segment must not have overlapping
      occupancy windows (deadlock/collision avoidance);
    * a greedy LPT dispatch minimising makespan.

  Pure planning compute; dispatches no real vehicle (G12 no-server-key). Electric AGV with
  regenerative braking (G8).

  Run:  bb --classpath 20-actors 20-actors/niyaku/methods/agv_transfer.clj"
  (:require [clojure.string :as str]))

(defn make-agv [& {:keys [v-max a-max length-m] :or {v-max 6.0 a-max 0.8 length-m 16.0}}]
  {:v-max v-max :a-max a-max :length-m length-m})

(defn- value-error [msg] (throw (ex-info msg {:type :value-error})))

(defn travel-time
  "Time-optimal travel time over a straight leg under a trapezoidal profile (triangular if too
  short to reach v-max)."
  [distance-m agv]
  (cond
    (< distance-m 0) (value-error "distance must be non-negative")
    (zero? distance-m) 0.0
    :else
    (let [a (:a-max agv) v (:v-max agv)
          d-to-vmax (/ (* v v) a)]               ; distance to accel to v-max then decel to 0
      (if (>= distance-m d-to-vmax)
        (let [t-ramp (/ v a) d-cruise (- distance-m d-to-vmax)]
          (+ (* 2.0 t-ramp) (/ d-cruise v)))
        (let [vp (Math/sqrt (* a distance-m))]   ; triangular peak velocity
          (/ (* 2.0 vp) a))))))

;; a reservation {:segment :agv-id :t-in :t-out}; a move {:move-id :distance-m}.
(defn reservation [segment agv-id t-in t-out]
  {:segment segment :agv-id agv-id :t-in (double t-in) :t-out (double t-out)})

(defn reservations-conflict?
  "True iff two reservations share a segment (different AGVs) and overlap in time. Touching at
  an endpoint (t_out == t_in) is NOT a conflict."
  [r1 r2]
  (if (or (not= (:segment r1) (:segment r2)) (= (:agv-id r1) (:agv-id r2)))
    false
    (and (< (:t-in r1) (:t-out r2)) (< (:t-in r2) (:t-out r1)))))

(defn find-conflicts
  "All conflicting index pairs [i j] (i<j) in a reservation vector."
  [reservations]
  (let [v (vec reservations) n (count v)]
    (vec (for [i (range n) j (range (inc i) n)
               :when (reservations-conflict? (v i) (v j))]
           [i j]))))

(defn move [move-id distance-m] {:move-id move-id :distance-m (double distance-m)})

(defn- argmin-first
  "Item of `coll` minimising (keyfn item); first one wins ties (matches Python's min)."
  [coll keyfn]
  (reduce (fn [best x] (if (< (keyfn x) (keyfn best)) x best)) (first coll) (rest coll)))

(defn dispatch
  "Greedy makespan-minimising assignment: each move (longest-first / LPT) goes to the AGV that
  frees up soonest. Returns {:assignment {agv [move-ids]} :finish-time {agv secs}}."
  [moves agv-ids agv]
  (when (empty? agv-ids) (value-error "need at least one AGV"))
  (let [assignment (atom (into {} (map (fn [a] [a []]) agv-ids)))
        finish (atom (into {} (map (fn [a] [a 0.0]) agv-ids)))]
    (doseq [mv (sort-by #(- (:distance-m %)) moves)]
      (let [a (argmin-first agv-ids #(@finish %))]
        (swap! assignment update a conj (:move-id mv))
        (swap! finish update a + (travel-time (:distance-m mv) agv))))
    {:assignment @assignment :finish-time @finish}))

(defn makespan [res]
  (if (seq (:finish-time res)) (reduce max (vals (:finish-time res))) 0.0))

(defn main [& _]
  (let [agv (make-agv)
        moves [(move "big" 300) (move "s1" 20) (move "s2" 20)]
        res (dispatch moves ["AGV1" "AGV2"] agv)]
    (println "niyaku AGV dispatch (LPT, makespan-balanced):")
    (doseq [[a ms] (sort-by key (:assignment res))]
      (println (format "  %s ← %s  (busy %.2fs)" a (str/join ", " ms)
                       (double (get (:finish-time res) a)))))
    (println (format "  makespan %.2fs" (makespan res)))))

(when (= *file* (System/getProperty "babashka.file"))
  (main))
